(feature) added date_passed & document_types to Legislations

kowal opened the pull request:
- Added `date_passed` and `document_types` (as tags) to `Legislations`.
- extracted some import helpers

you can re-import records via `bundle exec rake import:legislation`

<img width="1386" alt="Zrzut ekranu 2019-07-28 o 21 39 59" src="https://user-images.githubusercontent.com/10665/62011982-4c90d080-b180-11e9-8609-159ee15e3da7.png">

kowal commented on app/services/import_helpers.rb:
```diff
@@ -0,0 +1,17 @@
+    puts "Couldn't find Location with ISO: #{country_iso}"
+  end
+
+  def try_to_parse_date(date, expected_date_formats)
```
maybe rename it to `find_date_with_format`?

tsubik replied:
You can change it, for me both are good.

kowal replied:
I will leave it as it is.

tsubik commented on app/models/concerns/taggable.rb:
```diff
@@ -1,3 +1,25 @@
+# To tag your model with let's say 'keywords', you need to
```
Thanks for this :+1:

tsubik commented on app/models/document_type.rb:
```diff
@@ -0,0 +1,44 @@
+#
+
+class DocumentType < Tag
+  DOCUMENT_TYPES = [
```
Hmm, not sure if we want to restrict that list

tsubik replied:
Also, that list is funny now, I wouldn't pay too much attention to what is in the data currently. See there is document type `All`, they should clean this up, what they wanted to just tag Law with different types. Maybe we will have to extract some of those tags to be something else in the future (for example if application logic will depend on it, I suspect "Out of date" would be the first candidate)

kowal replied:
yeah I agree, this list will definitely change. For now I can remove the validation, but eventually, I think we will need some constraint on "document type". I guess this is a more precise term than, let's say "category".

tsubik replied:
Not sure, I think right now those are like categories, and definitely, that list needs cleanup from their side. I would just remove validation for now.

tsubik commented on app/services/import_helpers.rb:
```diff
@@ -0,0 +1,17 @@
+module ImportHelpers
```
Good idea

kowal commented on spec/controllers/admin/legislations_controller_spec.rb:
```diff
@@ -2,7 +2,16 @@
+
   let!(:legislation) { create(:legislation) }
+  let(:valid_attributes) {
+    attributes_for(:legislation).merge(
```
I couldn't find a way to include location by default in what's returned by `attributes_for` :/ then I found your approach from other models and used it here as well.

tsubik replied:
Yes, attributes_for are not working with associations, probably they have a good reason for that.

kowal commented:
@tsubik more updates:
- `date_passed`: added filter + default sort order + fix parsing some dates in importer
- better CSV output
- added more specs

tsubik commented on app/services/import_helpers.rb:
```diff
@@ -0,0 +1,17 @@
+module ImportHelpers
```
We can move it to `app/services/concerns`. I'm not sure if that would require to add new `concern` directory to autoload_paths in `config/application.rb`, like `config.autoload_paths += Dir[Rails.root.join("app", "services", "concerns")]`

kowal replied:
I think concerns are meant to encapsulate a piece of domain logic, here we have 2 simple helper methods related to different things.

Maybe I add 2 different modules, which could be used like this:

```ruby
def normalize_date
  Import::Date.try_to_parse(date)
end

def attributes
  { location: Import::Location.find(..) }
end

```

kowal replied:
Also, I'd move all import classes to `services/import` as well, so we'll end up with `Import::Companies.new.call` and so on.

tsubik replied:
Ok.

kowal replied:
done, please re-review
